src/api/workspace.rs

```diff
@@ -583,68 +578,72 @@ async fn batch_create_collab_handler(
+  let mut current_offset = 0;
+
   while let Some(item) = payload.next().await {
     if let Ok(bytes) = item {
-      match compress_type {
-        CompressionType::Brotli { buffer_size } => {
-          payload_buffer.extend_from_slice(&bytes);
-
-          // The client API uses a u32 value as the frame separator, which determines the size of each data frame.
-          // The length of a u32 is fixed at 4 bytes. It's important not to change the size (length) of the u32,
-          // unless you also make a corresponding update in the client API. Any mismatch in frame size handling
-          // between the client and server could lead to incorrect data processing or communication errors.
-          while payload_buffer.len() >= 4 {
-            let size = u32::from_be_bytes([
-              payload_buffer[0],
-              payload_buffer[1],
-              payload_buffer[2],
-              payload_buffer[3],
-            ]) as usize;
-
-            if payload_buffer.len() < 4 + size {
-              break;
-            }
-
-            let compressed_data = payload_buffer[4..4 + size].to_vec();
-            let decompress_data = decompress(compressed_data, buffer_size).await?;
-            let params = CollabParams::from_bytes(&decompress_data).map_err(|err| {
-              AppError::InvalidRequest(format!(
-                "Failed to parse CollabParams with brotli decompression data: {}",
-                err
-              ))
-            })?;
-            params.validate().map_err(AppError::from)?;
-            match params.check_encode_collab().await {
-              Ok(_) => collab_params_list.push(params),
-              Err(err) => error!("Failed to validate collab params: {:?}", err),
-            }
-
-            payload_buffer = payload_buffer[4 + size..].to_vec();
-          }
-        },
+      payload_buffer.extend_from_slice(&bytes);
+      while current_offset + 4 <= payload_buffer.len() {
+        // The length of the next frame is determined by the first 4 bytes
+        let size = u32::from_be_bytes([
+          payload_buffer[current_offset],
+          payload_buffer[current_offset + 1],
+          payload_buffer[current_offset + 2],
+          payload_buffer[current_offset + 3],
+        ]) as usize;
+
+        // Ensure there is enough data for the frame (4 bytes for size + `size` bytes for data)
+        if current_offset + 4 + size > payload_buffer.len() {
+          break;
+        }
+
+        // Collect the (offset, len) for the current frame (data starts at current_offset + 4)
+        offset_len_list.push((current_offset + 4, size));
+        current_offset += 4 + size;
       }
     }
   }
-  let duration = start_time.elapsed();
-  event!(
-    tracing::Level::DEBUG,
-    "end decompressing collab params list, time taken: {:?}",
-    duration
-  );
+  // Perform decompression and processing in a Rayon thread pool
+  let mut collab_params_list = tokio::task::spawn_blocking(move || {
```

Comment:
You're still spawning a thread in here.

Reply:
without using the spawn_blocking, it might block the tokio worker thread, causing the runtime to stall.